File: cpp/logagent_rpc_test.cpp
#include <stdio.h>
#include "logagent_rpc_test.h"
#include "logagent_rpc_cli.h"

namespace logagent {

	TestTool :: TestTool() {
	}

	TestTool :: ~TestTool() {
	}

	int TestTool :: critical( OptMap & ) {
		return -1;
	}

	int TestTool :: error( OptMap & ) {
		return -1;
	}

	int TestTool :: warning( OptMap & ) {
		return -1;
	}

	int TestTool :: info( OptMap & ) {
		return -1;
	}

	int TestTool :: debug( OptMap & ) {
		return -1;
	}

	int TestTool :: opreport( OptMap & ) {
		return -1;
	}

	int TestTool :: opquery( OptMap & ) {
		return -1;
	}

	int TestTool :: webreport( OptMap & ) {
		return -1;
	}

	int TestTool :: busireport( OptMap & ) {
		return -1;
	}


	TestToolImpl:: TestToolImpl() {
	}

	TestToolImpl:: ~TestToolImpl() {
	}

	int TestToolImpl :: critical( OptMap & opt_map ) {
		debugmsg req;

		Client cli;
		int ret = cli.critical( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: error( OptMap & opt_map ) {
		debugmsg req;

		Client cli;
		int ret = cli.error( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: warning( OptMap & opt_map ) {
		debugmsg req;

		Client cli;
		int ret = cli.warning( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: info( OptMap & opt_map ) {
		debugmsg req;

		Client cli;
		int ret = cli.info( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: debug( OptMap & opt_map ) {
		debugmsg req;

		Client cli;
		int ret = cli.debug( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: opreport( OptMap & opt_map ) {
		opmsg req;

		req.time = 1476678433;
		req.user = "david";
		req.action = "Logout";

		Client cli;
		int ret = cli.opreport( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: opquery( OptMap & opt_map ) {
		opqueryreq req;
		opqueryres res;

		req.user = "david";
		req.begintime = 1476678433;
		req.endtime = 1476678493;

		Client cli;
		int ret = cli.opquery( req, res );
		printf( "%s return %d\n", __func__, ret );
		for(std::list<opmsg>::iterator it = res.oplogs.begin();
				it != res.oplogs.end();
				++it) {
			printf( "\t%d %s %s %s %s\n", 
					it->time, 
					it->user.c_str(), 
					it->action.c_str(),
					it->args.c_str(),
					it->others.c_str() );
		}

		return 0;
	}

	int TestToolImpl :: webreport( OptMap & opt_map ) {
		webmsg req;

		Client cli;
		int ret = cli.webreport( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}

	int TestToolImpl :: busireport( OptMap & opt_map ) {
		busimsg req;

		Client cli;
		int ret = cli.busireport( req );
		printf( "%s return %d\n", __func__, ret );

		return 0;
	}


}

using namespace logagent;

void showUsage( const char * program )
{
	printf( "Usage:\n" );
	printf( "          %s [-c <config>] [-f <func>] [-h]\n", program );
	printf( "Options:\n" );
	printf( "          -c\tconfigure file of client\n" );
	printf( "          -f\trpc method or function\n" );
	printf( "          -h\tshow help\n" );
	printf( "Examples:\n");

	TestTool::Name2Func_t * name2func = TestTool::GetName2Func();

	for( int i = 0; ; i++ ) {
		TestTool::Name2Func_t * iter = &( name2func[i] );

		if( NULL == iter->name ) break;

		printf( "          %s -c logagent_rpc_cli.conf -f %s %s\n", program, iter->name, iter->usage );
	}

	exit( 0 );
}

int main( int argc, char * argv[] )
{
	const char * func = NULL;
	const char * config = NULL;

	for( int i = 1; i < argc - 1; i++ ) {
		if( 0 == strcmp( argv[i], "-c" ) ) {
			config = argv[ ++i ];
		}
		if( 0 == strcmp( argv[i], "-f" ) ) {
			func = argv[ ++i ];
		}
		if( 0 == strcmp( argv[i], "-h" ) ) {
			showUsage( argv[0] );
		}
	}

	if( NULL == func ) showUsage( argv[0] );

	if( NULL != config ) Client::Init( config );

	TestTool::Name2Func_t * target = NULL;

	TestTool::Name2Func_t * name2func = TestTool::GetName2Func();

	for( int i = 0; i < 100; i++ ) {
		TestTool::Name2Func_t * iter = &( name2func[i] );

		if( NULL == iter->name ) break;

		if( 0 == strcasecmp( func, iter->name ) ) {
			target = iter;
			break;
		}
	}

	if( NULL == target ) showUsage( argv[0] );

	OptMap opt_map( target->opt_string );

	if( ! opt_map.Parse( argc, argv ) ) showUsage( argv[0] );

	TestTool::ToolFunc_t targefunc = target->func;

	TestToolImpl tool;

	if( 0 != ( tool.*targefunc ) ( opt_map ) ) showUsage( argv[0] );

	return 0;
}

File: python/rpc_test/logagent_rpc_test.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

import os
import time
import importlib
import multiprocessing.dummy

class RPCTest:
	def __init__(self):
		rpc_test_name = os.path.splitext(os.path.basename(__file__))[0]
		service_name = rpc_test_name[:-9]
		self.rpc_proto = None
		self.cli = self._get_rpc_cli().Client(service=service_name, caller=rpc_test_name)

	def _get_rpc_cli(self):
		absolute_path = os.path.dirname(os.path.abspath(os.path.join(__file__, '..')))
		for python_path in os.getenv('PYTHONPATH').split(':'):
			if absolute_path.startswith(python_path):
				rootpath = os.path.relpath(absolute_path, python_path).replace(os.path.sep, '.')
				self.rpc_proto = importlib.import_module('.rpc_proto.logagent_rpc_proto', rootpath)
				return importlib.import_module('.rpc_cli', rootpath)
		else:
			raise IOError('%s is not in PYTHONPATH' % (__file__))

	def run(self):
		self.test_critical()
		self.test_error()
		self.test_warning()
		self.test_info()
		self.test_debug()
		self.test_opreport()
		self.test_opquery()
		self.test_webreport()
		self.test_busireport()
		#self.test_multithread()

	def test_critical(self):
		ret = self.cli.critical('%s %s', 'test', 'critical')
		print 'critical ret %s' % (ret)

	def test_error(self):
		ret = self.cli.error('%s %s', 'test', 'error')
		print 'error ret %s' % (ret)

	def test_warning(self):
		ret = self.cli.warning('%s %s', 'test', 'warning')
		print 'warning ret %s' % (ret)

	def test_info(self):
		ret = self.cli.info('%s %s', 'test', 'info')
		print 'info ret %s' % (ret)

	def test_debug(self):
		ret = self.cli.debug('%s %s', 'test', 'debug')
		print 'debug ret %s' % (ret)

	def test_opreport(self):
		req = self.rpc_proto.opmsg()
		req.time = 1476678433
		req.user = 'david'
		req.action = 'Logout'
		ret = self.cli.opreport(req)
		print 'opreport ret %s' % (ret)

	def test_opquery(self):
		req = self.rpc_proto.opqueryreq()
		req.user = 'david'
		req.begintime = 1476678433
		req.endtime = 1476678493
		ret = self.cli.opquery(req)
		print 'opquery ret %d oplogs:' % (len(ret.oplogs))
		for oplog in ret.oplogs:
			print '\t', oplog.time, oplog.user, oplog.action, oplog.args, oplog.others

	def test_webreport(self):
		req = self.rpc_proto.webmsg()
		ret = self.cli.webreport(req)
		print 'webreport ret %s' % (ret)

	def test_busireport(self):
		req = self.rpc_proto.busimsg()
		ret = self.cli.busireport(req)
		print 'busireport ret %s' % (ret)

	def test_multithread(self):
		thread_sum = 5
		thread_pool = []

		for i in xrange(thread_sum):
			t = multiprocessing.dummy.DummyProcess(
				target = self.thread_logic,
				args = (i, self.cli))
			thread_pool.append(t)
			t.start()

		for t in thread_pool:
			t.join()

	def thread_logic(self, tid, client):
		while 1:
			client.info('thread%d test info', tid)
			time.sleep((tid+1)*0.01)

if __name__ == '__main__':
	RPCTest().run()

